src/chi/javascript/expansion-panel.js

```diff
@@ -0,0 +1,405 @@
+      }
+    };
+    this._actions.inactive = this._actions.pending;
+  }
+
+  _clickHandler (e) {
+    if (!e.target || !e.target.dataset || !e.target.dataset.chiEpAction) {
+      return;
+    }
+    this.execute(e.target.dataset.chiEpAction);
+  }
+
+  execute (action) {
+    if (this._actions[action]) {
+      this._actions[action]();
+    }
+  }
+
+  setState (stateName) {
+    const newState = getStateFromName(stateName);
+    if (newState && newState !== this._state) {
+      const oldState = this._state;
+      this._state = newState;
+      if (this._config.animated) {
+        this.animateStateChange(oldState, newState);
+      } else {
+        Util.addClass(this._elem, newState.CLASS);
+        Util.removeClass(this._elem, oldState.CLASS);
+      }
+      this._stateChangeListeners.forEach(
+        listener => listener(newState, oldState, this)
+      );
+    }
+  }
+
+  addStateChangeListener (listener) {
+    this._stateChangeListeners.push(listener);
+  }
+
+  removeStateChangeListener (listener) {
+    const index = this._stateChangeListeners.indexOf(listener);
+    if (index > -1) {
+      this._stateChangeListeners.splice(index, 1);
+    }
+  }
+
+  animateStateChange (oldState, newState) {
+
+    if (this._transitioning) {
+      window.clearTimeout(this.animationEndTimeoutId);
+      this.animationEnd();
+    }
+
+    this._transitioning = true;
+
+    const
+      viewIn = {found: false, elem: null, height: 0, oldStyle: {}},
+      viewOut = {found: false, elem: null, height: 0, oldStyle: {}},
+      wrapper = {found: false, elem: null, height: 0, oldStyle: {}};
+
+    const updateStyle = function (elem, property, value, writeOnlyIfNotEmpty) {
+      elem.oldStyle[property] = elem.elem.style[property];
+      if (writeOnlyIfNotEmpty && !elem.elem.style[property]) {
+        elem.elem.style[property] = value;
+      } else if (!writeOnlyIfNotEmpty) {
+        elem.elem.style[property] = value;
+      }
+    };
+    const restoreStyle = function (elem) {
+      Object.getOwnPropertyNames(elem.oldStyle).forEach(function (key) {
+        elem.elem.style[key] = elem.oldStyle[key];
+      });
+    };
+
+    const prepare = function () {
+
+      Util.findAndApply(this._elem, CLASS_CONTENT_VIEW, function (elem) {
+        if (Util.hasClass(elem, oldState.CLASS_TARGET)) {
+          viewOut.found = true;
+          viewOut.elem = elem;
+          updateStyle(viewOut, 'opacity', '1', true);
+
+          viewOut.height = window.getComputedStyle(elem).height;
+        } else if (Util.hasClass(elem, newState.CLASS_TARGET)) {
+          viewIn.found = true;
+          viewIn.elem = elem;
+          updateStyle(viewIn, 'opacity', '0');
+          updateStyle(viewIn, 'position', 'absolute');
+          updateStyle(viewIn, 'top', '0');
+          updateStyle(viewIn, 'left', '0');
+          updateStyle(viewIn, 'right', '0');
+          updateStyle(viewIn, 'background-color', 'white');
```

Comment:
the white background used in the transition isn't going to render well if the panel background is set to transparent or colored some shade of grey